LFS-638: Data tables improvements. Includes subtasks LFS-607, LFS-555, LFS-606, LFS-609

veronikaslc opened the pull request:
[LFS-607](https://phenotips.atlassian.net/browse/LFS-607): In all grids replace current Identifiers with user friendly formatted identifiers
[LFS-555](https://phenotips.atlassian.net/browse/LFS-555): In all data tables, show full identifiers (parents included if any) for subjects and forms and reorganize columns for better readability
[LFS-606](https://phenotips.atlassian.net/browse/LFS-606): Switch "Created by" and "Created on" columns in all data grids across LFS
[LFS-609](https://phenotips.atlassian.net/browse/LFS-609): Add type column to Subjects grid

Commit:
LFS-607: In all grids replace current Identifiers with user friendly formatted identifiers

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/QuickSearchResults.jsx:
```diff
@@ -33,6 +34,31 @@ const LFS_QUERY_MATCH_TEXT_KEY = "text";
 const LFS_QUERY_MATCH_NOTES_KEY = "inNotes";
 
+// Display the identifier column of the result item wrt item primaryType
+export function MatchIdentifier(row) {
```
Since this is used in so many places that are not related to QuickSearch, this method should be put in a file of its own.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/QuickSearchResults.jsx:
```diff
@@ -33,6 +34,31 @@ const LFS_QUERY_MATCH_TEXT_KEY = "text";
+// Display the identifier column of the result item wrt item primaryType
+export function MatchIdentifier(row) {
+    const anchorPath = row[LFS_QUERY_MATCH_KEY]?[LFS_QUERY_MATCH_PATH_KEY] : '';
+    switch (row["jcr:primaryType"]) {
```
This will need to be made extensible, but not right now.

marta- reviewed (changes requested):
Please rename the `MatchIdentifier` to `EntityIdentifier` and pull it out it in its own file. Also, please use this new identifier format in the Quick search dropdown as well.

marta- commented on modules/homepage/src/main/frontend/src/themePage/QuickSearchResults.jsx:
```diff
@@ -33,6 +34,31 @@ const LFS_QUERY_MATCH_TEXT_KEY = "text";
+        let questionnaire = (row.questionnaire?.title?.concat(' ') || '')
+          + (row["jcr:primaryType"]?.replace(/lfs:/,"") || '');
+        let subjectHierarchy = row.subject ? getHierarchy(row.subject, Link, (node) => ({to: "/content.html" + node["@path"], target :"_blank"})) : '';
+        let formpath = `/content.html/admin${row["@path"]}#${anchorPath}`;
```
Wrong path

```suggestion
        let formpath = `/content.html/${row["@path"]}#${anchorPath}`;
```

sashaandjic commented:
I think we should drop "Patient", "Tumor" and "Data Form" from ids.
 
**Patient John/Tumor Lung: Radiotherapy Data Form** would become **John/Lung: RADIOTHERAPY**
**Patient John/Tumor Liver** would become **John/Liver**
**Patient John/tumor Liver/Tumor region Slice1** would become **John/Liver/Slice1**

![image](https://user-images.githubusercontent.com/23243244/98852694-0a97a680-2426-11eb-96b5-da3199279710.png)
![image](https://user-images.githubusercontent.com/23243244/98852710-0ff4f100-2426-11eb-9a0f-faf500b9f458.png)

sashaandjic commented:
The form links are not working 
![image](https://user-images.githubusercontent.com/23243244/98854099-213efd00-2428-11eb-9a1a-ca19d3cafa05.png)

Commit:
LFS-607: In all grids replace current Identifiers with user friendly formatted identifiers
Separating EntityIdentifier into new file; Edited Quick search results page

sashaandjic commented:
I also noticed that subjects are now opening in a new tab.

Commit:
LFS-607: In all grids replace current Identifiers with user friendly formatted identifiers
Removed subject types from ids

veronikaslc commented:
All comments are addressed

sashaandjic commented:
Looking really good! Subjects are still opening in the new tab.
![image](https://user-images.githubusercontent.com/23243244/98861283-02923380-2433-11eb-95df-37536a36de98.png)
![image](https://user-images.githubusercontent.com/23243244/98861901-05415880-2434-11eb-8bfe-b50aae6586bc.png)

veronikaslc commented:
That's by design in previous issue about subjects.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
> That's by design in previous issue about subjects LFS-650. It stated in the issue description:
> "Add links to the subject ids in the hierarchy. Links should open in a new tab"

I was not aware of this.

marta- commented:
> That's by design in previous issue about subjects LFS-650. It stated in the issue description:
> "Add links to the subject ids in the hierarchy. Links should open in a new tab"

This was specified only for form pages (LFS-650). The reason was so that the user doesn't accidentally navigate away from the form, losing their unsaved data. LFS-653, which targets subject pages, did not mention opening in a new tab.

I agree with Sasha that in most cases there's no need for opening subject links in a new tab.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -51,11 +51,11 @@ let createQueryURL = (query, type) => {
 
 // Recursive function to get a flat list of parents
-export function getHierarchy (node, RenderComponent, propsCreator) {
+export function getHierarchy (node, RenderComponent, propsCreator, includeSubjectLabel) {
```
I think the code would be easier to understand if we rename this:

```suggestion
export function getHierarchy (node, RenderComponent, propsCreator, includeSubjectType) {
```

Commit:
LFS-653: Subject links on chart pages are not functional
Removed subjects opening on new tab from subjects hierarchy links

Commit:
LFS-607: In all grids replace current Identifiers with user friendly formatted identifiers
Removed subjects opening on new tab from subjects hierarchy links

Commit:
LFS-607: In all grids replace current Identifiers with user friendly formatted identifiers
Better name for variable

Commit:
LFS-555: Remove columns showing redundant information from data tables
Forms page: removed identifier and swaped subject id and questionnaire columns
Dashboard: left only subject identifier and created on/by columns

Commit:
LFS-606: Switch Created by and Created on columns in all data grids across LFS

Commit:
LFS-609: Add type column to Subjects grid

sashaandjic commented:
retested. all good!

veronikaslc commented:
As required in LFS-555:
"Forms page: remove identifier and swap subject id and questionnaire..."

sashaandjic commented:
I think there was a misunderstanding of the LFS-555 details. 

By removing Identifier, Marta meant remove the old user **un**friendly identifier (not the newly created user friendly one; this was already done in LFS-607).

Data tables should have the following columns: 
- Identifier (user friendly as specified in LFS-607)
- Questionnaire (on Forms data table only to allow sorting/filtering later on; redundant in COMPLETED data tables)
- Subject (user friendly e.g. Mary/Liver instead of just Liver)
- Subject Type

Dashboard INCOMPLETE:
- Identifier 
- Created on
- Created by

Commit:
LFS-555: Remove columns showing redundant information from data tables
Added identifier columns back to Forms and Dash board table; Fixed delete button regression

veronikaslc commented:
Added identifier columns back to Forms and Dash board table cause most likely LFS-555 is not defined precisely.

marta- commented:
> As required in LFS-555:
> "Forms page: remove identifier and swap subject id and questionnaire...
> Dashboard: only display subject identifier and created on/by columns"

Should have been more precise: I lost track of the fact that full identifiers contain links to each subject in the hierarchy and was envisioning the subject id as one solid link to the form. Looking at the current implementation, I don't think it looks too cluttered with the full form name, so we can keep it.

However, now the Subject column on both Dashboard and Forms looks awkward (e.g. 3 Tumor - only the last child in the hierarchy with Subject type _after_ the id). I would remove it from the dashboard, and on Forms I would display it as full id including subject type (e.g. Patient 123 / Tumor 3).

@sashaandjic @veronikaslc thoughts?

sashaandjic commented:
I think this would be a good look for the Dashboard:
![image](https://user-images.githubusercontent.com/23243244/98950392-f061d500-24c6-11eb-9754-99adb68b6a60.png)

marta- commented:
> I think this would be a good look for the Dashboard:

Agreed

sashaandjic commented:
Here is the final look
![image](https://user-images.githubusercontent.com/23243244/98950822-7aaa3900-24c7-11eb-9824-33b7f2f0e9e4.png)

marta- commented:
Not sure if the "undefined" issue in delete dialogs on the Subjects page was introduced here or not (screenshot attached). However the Delete dialog should display the full id (parents included if any, no lonk).

![image](https://user-images.githubusercontent.com/651980/98974310-c3bbb680-24e2-11eb-8c05-88fe0fd8c33e.png)

sashaandjic commented:
> Not sure if the "undefined" issue in delete dialogs on the Subjects page was introduced here or not (screenshot attached).

I'll check.

sashaandjic commented:
> Not sure if the "undefined" issue in delete dialogs on the Subjects page was introduced here or not (screenshot attached). However the Delete dialog should display the full id (parents included if any, no link).
> 
> ![image](https://user-images.githubusercontent.com/651980/98974310-c3bbb680-24e2-11eb-8c05-88fe0fd8c33e.png)

Probably introduced in this PR. Cannot reproduce it in the DEV branch.

Commit:
LFS-555: Remove columns showing redundant information from data tables
Refactoring for better UI

marta- commented:
Subject charts don't load. Blank screen and console error:
```
react-dom.development.js:23965 Uncaught Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `Subjects`.
    at createFiberFromTypeAndProps (react-dom.development.js:23965)
    at createFiberFromElement (react-dom.development.js:23988)
    at reconcileSingleElement (react-dom.development.js:14233)
    at reconcileChildFibers (react-dom.development.js:14293)
    at reconcileChildren (react-dom.development.js:16769)
    at updateFunctionComponent (react-dom.development.js:17053)
    at beginWork (react-dom.development.js:18610)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
```

marta- commented:
I think we should imention the subject types in the delete dialogs, like this:
![image](https://user-images.githubusercontent.com/651980/99084466-fa520980-2594-11eb-9af6-a6c49b0ce301.png)

veronikaslc committed:
LFS-555: In all data tables, show full identifiers (parents included if any) for subjects and forms and reorganize columns for better readability

Fixed Subject import

sashaandjic reviewed (approved):
Looks great!
Now that grids are no longer cluttered, the row appear to be too high. However this is a minor complaint and probably out of scope of this PR.
![image](https://user-images.githubusercontent.com/23243244/99126584-79b2fd80-25d4-11eb-9116-43e6582fcd1a.png)
I do believe Marta wanted to see "Delete **Tumor region** Joe/Liver/Slice?"

![image](https://user-images.githubusercontent.com/23243244/99125667-8df5fb00-25d2-11eb-9714-c278537cbc39.png)
![image](https://user-images.githubusercontent.com/23243244/99125682-964e3600-25d2-11eb-864d-ad31c2e5452d.png)
![image](https://user-images.githubusercontent.com/23243244/99125699-a1a16180-25d2-11eb-947d-bd6cb3939909.png)

veronikaslc committed:
LFS-555: In all data tables, show full identifiers (parents included if any) for subjects and forms

Introduced entry label to the Delete dialog to be used in header

veronikaslc committed:
LFS-555: In all data tables, show full identifiers (parents included if any) for subjects and forms and reorganize columns for better readability

Set entry label for Subjects to be used in the Delete dialog title

veronikaslc committed:
LFS-555: In all data tables, show full identifiers (parents included if any) for subjects and forms and reorganize columns for better readability

Typo fix

sashaandjic commented:
Subject page is now broken. After creation of a new patient, the subject page went blank. Same with forms. I created demographic form, picked the patient and the screen went blank.

Commit:
LFS-555: In all data tables, show full identifiers (parents included if any) for subjects and forms and reorganize columns for better readability
Proper typo fix

sashaandjic commented:
retested; all good!
Only one very minor complaint. In the delete dialog the subject type is only mentioned in the title, but not in the detail
![image](https://user-images.githubusercontent.com/23243244/99215541-3b1e6e00-27a1-11eb-8e62-53d03576a073.png)
![image](https://user-images.githubusercontent.com/23243244/99215549-41ace580-27a1-11eb-9302-f8937e71df2f.png)

Also, the build even though built successfully, there were few errors in the log.
[LFS-607 build log.txt](https://github.com/ccmbioinfo/lfs/files/5544269/LFS-607.build.log.txt)

Commit:
LFS-555: In all data tables, show full identifiers (parents included if any) for subjects and forms and reorganize columns for better readability
Added type label to the delete dialog message

marta- reviewed (approved):
Reviewed and tested

marta- commented:
> One very minor complaint: In the delete dialog the subject type is only mentioned in the title, but not in the detail

Valid point, but I think that's ok for now, it's clear. I have some general improvements in mind for the wording in delete dialogs, and we'll cover this issue then.
